### DISCUSSION TITLE
MissionControl 0.7.0

### DISCUSSION THREAD

**Discussion by ndeadly** [ndeadly/MissionControl#446] [General]
<p align="left">
<img alt="GitHub Releases" src="https://img.shields.io/github/downloads/ndeadly/MissionControl/v0.7.0/total">
</p>

This is the fourteenth official release of Mission Control.

This adds full support for HOS 14.1.0, **motion controls** for Dualshock4, Dualsense and Wii MotionPlus* controllers, and support for the g910 Wireless Bluetooth controller. See below for the full changelog.

_*Note: regular Wiimote controllers without the MotionPlus extension _may_ work in some games, but only come equiped with an accelerometer and lack the gyroscope sensor required for full "sixaxis" motion controls._

A big thanks to @Banz99 for testing and giving feedback on the Wii MotionPlus controller, and to @Hydr8gon for porting and verifying Dualshock4 motion support to Dualsense controllers.

New users landing on this release page should first check out the readme on the [main project page](https://github.com/ndeadly/MissionControl#readme) for the official project documentation. There you can find installation and usage instructions along with an FAQ section that will answer most of your questions.

***Atmosphère >= 1.3.1 is required to run this release on firmware 14.1.0.***

### Notice for users of Xbox One controllers
If you have updated to the new bluetooth LE controller firmware and lost support for your controller, it is now possible to downgrade back to a compatible firmware using an Xbox One console. See [here ](https://answers.microsoft.com/en-us/xbox/forum/all/revertrollbackdowngrade-xbox-controller-firmware/25f07169-3244-4fe0-838c-949951b7c109?auth=1) for details.

@HaruVM has pointed out to me that it's now also possible to downgrade the controller firmware using a PC running the [Xbox Accessories](https://www.microsoft.com/en-us/p/xbox-accessories/9nblggh30xj3?activetab=pivot:overviewtab) app.

> Open xbox accessories app then hit windows + R (run) and input this command:
> 
> xboxaccessories:\\\firmwareupdate?legacyDowngrade=true
> 
> It should then give you option to revert

### Changelog:
- Added bluetooth ips patches for 14.0.0.
- Added support for g910 wireless bluetooth controller.
- Support for motion controls was added for Dualshock4, Dualsense, and Wii MotionPlus controllers.
- Wii controller handling was refactored for accuracy and completeness.
- Added mechanisms for executing code in a worker thread and receiving asynchronous responses from controllers.
- Minor bugfixes and improvements

<hr /><em>This discussion was created from the release <a href='https://github.com/ndeadly/MissionControl/releases/tag/v0.7.0'>MissionControl 0.7.0</a>.</em>

**Comment by jacoghi** [comment]
Wow... No comments whatsoever... Motion controls on the DS4? That's simply awesome, I'll get down to testing it right away, are there any known limitations? If not, this now basically replaces entirely any need for 8bitdo adapters, which is unthinkable of. Awesome, congrats ndeadly and Hydr8gon, amazing work as always, man

**Reply by ndeadly** [reply]
Cheers! No hard limitations that I know of for DS4/Dualsense. I haven't really tried it out during gameplay, but I made data captures of the calibrated motion data as the console sees it and things look as close to 1:1 with a Pro Controller as you could expect given different IMU hardware.

Only potential issues I can think of are regarding sensitivty. There is a command that can be sent to the controllers to request different sensitivity levels of the accelerometer and gyro. I don't know how much (if at all) this is actually used in practice, but currently I don't attempt to handle this scenario and always use the default sensitivity values. If it turns out to be an issue with some games, this could easily be resolved be taking the sensitivy setting into account and introducing a scaling factor.
The other thing relates to the data rate of the controller. The switch IMU runs at a higher rate than the controller sends data, and can send up to 3 motion samples per input packet, with a counter to indicate how many new samples arrived. These samples are expected to be spaced 5ms apart. If a third-party controller is sending data at a different rate, eg spacing of <5ms or not divisible by 5, then we are not able to accurately indicate the spacing between the samples. The actual data will be correct, but the time delta will be different. If you're familiar with calculus you will know that this time delta is important when eg. integrating velocity to find position. We're lucky enough that the Sony controllers can be run at more or less the same rate as the Switch Pro controller, but any deviations from this can result in rotation angles that are not 1:1 with the controller movements even though the acceleration and angular velocity reported may be correct. This is unlikely to be a dealbreaker for anyone, as it is again just a sensitivity issue, but may create an inconsistent feel between different controllers. We could potentially solve this by interpolating between data points if necessary.

**Reply by jacoghi** [reply]
If I'm not mistaken, those samples at least on the DS4 are normally sent 5 ~ 8 ms apart, correct? So we should be good when it comes to that. I actually tested the functionality yesterday and it works beautifully, man, I couldn't tell the difference between using MC and the 8bitdo adapter, so I'm ditching it completely now since your solution is so much more beautiful. Again, big shoutout and appreciation for this, man.

**Reply by ndeadly** [reply]
There are a number of different reporting rates allowed, but the one I'm using by default is 125Hz. That should be about 8ms between samples in practice, yeah. Not identical, but I guess it's good enough for now by the sounds of things. I may make an effort to improve accuracy in the future if any tangible differences arise. Thanks for the kind words, glad you're enjoying it.

**Comment by n0ah119** [comment]
Great job!!!Motion controls support is a highlight that far exceeds expectations. 
May I ask what is the progress of support on xbox series s/x BLE stack in 14.1 ? 
 Congrats again 👍

**Reply by ndeadly** [reply]
No progress since the last reports if you've been following along. I have a proof of concept working, but still not sure on how to get the controllers integrated with the rest of the system. Some of the work on motion controls lays a foundation for some of the things I need to be able to do for BLE, so with that out of the way I can begin to look at BLE support more seriously. There are still a lot of unknowns though, so hard to say how far along it is.

**Comment by KAMELCOLTD** [comment]
you did a very good job ndeadly adding "motion" , i don't understand 100% but dose the g910  had motion controller ? or !!
since i had this controller since 2021 and it works perfectly with mission controller is there anything else new for it !!

♥ thank you for your hard working

**Reply by ndeadly** [reply]
I don't know much about the g910 controller, but it seemed to be getting picked up as an ipega controller. The user that requested told me it was working except for L3/R3 so I modified the ipega handler to map those properly. Maybe yours had this problem too but you never tried those buttons, or maybe this user simply had a different firmware with a slightly different layout 🤷‍♂️

**Reply by KAMELCOLTD** [reply]
nope mine works good , everything is good 
there is only 1 problem after pairing the controller it work perfectly .
but next time when i turn it on it won't paired until i open the switch controller ui , i have to do that every time  .
it won't pair if i don't open the grip change ui  lol , strange but it's ok

**Reply by ndeadly** [reply]
Again, not really familiar with this controller so I'm not sure how it's supposed to work, but if you're having to go into the change grip UI to connect it, it sounds like you're starting the controller in pairing mode. Most controllers only require you to hit a button for them to reconnect to the last device they were paired with. I don't think it's an issue with Mission Control or HOS.

**Comment by JoshuaTLOZ** [comment]


**Reply by ndeadly** [reply]
Had a few reports of this now, but I'm not exactly sure about the cause. 0.7.0 made changes to change what would normally have been a system crash, into just a disconnection of the controller that caused the issue. The downside is that we don't get an error report when this happens.

Take a look at [these suggestions](https://github.com/ndeadly/MissionControl/issues/449#issuecomment-1090993951) and see if any of them work for you. I have my suspicions that it might be something to do with the virtual controller memory files becoming corrupted or something, so deleting the controllers directory is what I would wager on being most likely to help.

**Reply by JoshuaTLOZ** [reply]
I don't know why but I fully recharged the DS4 and now works. Just did that.

**Reply by ndeadly** [reply]
Heh, maybe these issues are all just low battery problems then. Didn't even consider that, but would make sense given your symptoms.

**Comment by Account-is-Deleted** [comment]
Hello,
I am new here,
Just curious as to whether haptics/vibrations are currently supported and if not, is there any way to get them to work?
Thanks a lot,
Your tool is awesome!

**Reply by ndeadly** [reply]
Are you referring to a specific controller here? Rumble is already supported for controllers that I know the output packet format for. This includes most of the major console controllers. It can be added for other more obscure controllers that support it, but someone would have to inform me of the data format through packet captures or otherwise.

Since you mention haptics, if you're referring specifically to Dualsense controllers I can't support their haptic feedback currently.

**Reply by Account-is-Deleted** [reply]
Sorry I was not specific, I am talking about rumble on both the dualsense and the xbox one controllers, I can get the controllers connected but no matter what I do, I cannot get them to rumble.
Thanks for replying so quickly btw, I really appreciate it :D

**Reply by ndeadly** [reply]
Basic rumble should work for both of those. Keep in mind that it is not 1:1 with the HD rumble of the switch controllers, so it's posible that some low amplitude/short duration vibration patterns are lost.
Make sure that you don't have controller vibration switched off in the system settings. Also check to make sure you don't have a config file (`sdmc:/config/MissionControl/missioncontrol.ini`) with `enable_rumble=false` set. This shouldn't be present by default, but if you've installed from an unofficial source at any point they may have added the config

**Reply by Account-is-Deleted** [reply]
Update! Turns out rumble does not work on the controller that I am using.
It now works with the dualsense.
Thanks for your Help :D

**Comment by SpicyMchagis** [comment]
Hey I love this app! Is it working on 14.1.1?

**Reply by ndeadly** [reply]
Thanks :) Should be, 14.1.1 only made minor changes and doesn't require an Atmosphere update as far as I know.

**Reply by SpicyMchagis** [reply]
K awesome, thank you! 

Edit: my DS4 is not being picked up by the switch. Just updated everything on my switch tonight. Figured I'd give a heads up. Maybe I'm doing something wrong. 

Edit 2: forgot to reboot! Picked it up right away!

**Reply by demontaim** [reply]
How you got it working? I'm on CFW 14.1.0 before upgrading my sd card eveything worked, when I switched my sd card for the new one with everything installed and with the latest release it seems like connecting in a light blue color not in the navy blue one. The app in the homebrew that says how many devices you got connected does recognize it but I cant use it, If I switch my sd to the previous one it works.

**Reply by demontaim** [reply]
I've tried uninstalling eveything, installing it by the hb shop, deleting my controllers database, reseting the ds4 but I don't now what am I doing wrong.

**Comment by nathangusman** [comment]
any predictions for motion control in dualshock 4?

**Reply by ndeadly** [reply]
> any predictions for motion control in dualshock 4?

Predictions?

> This adds full support for HOS 14.1.0, motion controls for **Dualshock4**, Dualsense and Wii MotionPlus* controllers

It's supported in this release already, if that's what you mean.

**Reply by nathangusman** [reply]
sorry, it's that on the github home screen it says that motion control is not being used

**Reply by nathangusman** [reply]
[general] These are general settings for mission control features.

enable_rumble Enables/disables rumble support for unofficial controllers.
enable_motion Enables/disables motion controls support. Not currently used

**Reply by ndeadly** [reply]
Oh, that needs to be updated in the readme then, since it's now in use. In any case that's just an option to globally switch 3rd party motion control mapping on/off and has nothing to do with whether the support is actually implemented or not.

**Comment by Hulir** [comment]
Hello, nice to meet you, I have a problem with my GXT 545 control that no longer synchronizes, I remember more or less that it was between 11.0.1 and 12.0.1 it stopped synchronizing. when I was with sx os it worked for me and when I changed to atmosphere it also worked for me but after updating it stopped synchronizing.

**Reply by ndeadly** [reply]
This controller has never been officially supported, it must identify the same as some other controller if it was working. Some big changes were made to the bluetooth service in the 12.0.0 update (and also the mission control update to support it) so this sounds plausible. What happens now when you try to pair it? Do you get and pairing notification on the screen or just nothing?

**Reply by Hulir** [reply]
Simply nothing, I connect the wireless USB of the control and nothing happens, as I said before it pairs me without any problem, it even came out as a light blue opaque control in synchronization when I used it. I don't think I installed the mission control wrong since the ps5 control works for me

**Reply by Hulir** [reply]
The controller works on ps3 and pc, do you think mission control will detect it as a ps3 controller?

**Reply by ndeadly** [reply]
Wireless USB? If there's any USB adapter involved then it's not talking directly to the switch Bluetooth chip, and therefore not running through mission control. It may not even be communicating via Bluetooth at all.  Are you sure you weren't actually connecting via this adapter with sys-con before?

**Reply by ndeadly** [reply]
> The controller works on ps3 and pc, do you think mission control will detect it as a ps3 controller?

No, mission control doesn't support ps3 controllers.

**Reply by Hulir** [reply]
I would have to see, that I remember I did not install a program with that name, I will see in the old files that I have on the PC, and will it still be compatible with the new versions of the hack? Thanks for helping.

**Reply by ndeadly** [reply]
Put it this way - if you rely on a USB dongle for a connection then you weren't running through mission control, as you are effectively using a USB controller. Unless the controller has native Switch support then you would necessarily be running a USB controller sysmodule such as sys-con. As far as I know there aren't any others. As for whether it work with the latest firmware I couldn't tell you, as I don't use it. Try it and see.

**Reply by Hulir** [reply]
![IMG_20220506_121645](https://user-images.githubusercontent.com/58637460/167172365-e523ff52-9b59-4143-8d11-7882fc0a1449.jpg)
Great it works for me with the application you told me, I saw that if I had it before I'm sorry for bothering you, really thanks now it works

**Reply by ndeadly** [reply]
All good, happy you got it figured out :)

**Comment by sandroddr** [comment]
Can you use wii controller as a motion joy con to play games like mario party, or do you have prediction? When I put the wii controller it only recognizes it as a pro controller. Thanks to all developers for their work.

**Reply by ndeadly** [reply]
No, not currently. It's a feature that has been looked into already a while ago, but it's not ready for release yet.

**Comment by xgod-keith** [comment]
Hi!
Recently updated to 14.1.1 and my mission control to 0.7.0
(i have a dualshock 4 v2)
so, the licensed games from nintendo, doesn't works, i've tried with megaman zero collection and works fine, but i tried with BOTW, Mario collection and mario odyssey, for my bad luck not work, the buttons are dead. (an exception the touchpad-capture button works)
what i can do?
what is my issue? tnx

**Reply by ndeadly** [reply]
That's strange. So the controller is connecting but only the touchpad button works? Is the controller LED colour assigned correctly (blue for player 1)? 

Try running [this ](https://github.com/ndeadly/MissionControl/files/7628384/hid_report_viewer.zip) with the controller connected and make a screenshot so I can get an idea of the report format the controller is sending.

**Comment by fahoodxx** [comment]
hi I tried to downgrade my series x controller and it's
![image](https://user-images.githubusercontent.com/66586605/169158737-0b6ec906-b9b1-4bb4-827e-4d54594d5699.png)
 not working it does not give the option to downgrade

**Reply by ndeadly** [reply]
The Series X controller was released with a Bluetooth LE firmware. There is no legacy firmware to downgrade to. The downgrade instructions are for Xbox One controllers that have been updated to the LE firmware. Unfortunately for the new controllers you will have to wait until LE support is complete.

**Comment by Codecor1** [comment]
hey, I have a doubt, I have filmware 14.1.2 and atmosphere 1.3.2 and the page says that the version of atmosphere must be "Atmosphère >= 1.3.1 is required to run this release on firmware 14.1.0.", but it still won't let me connect my ps4 joystick even though the page says What is supported, could you help me? Thanks in advance (sorry if my inglish is bad)
PS: I tried to restart the joystick and connect it but there is no way for it to connect (same problem with my xbox one joysticks)

**Reply by ndeadly** [reply]
Could you describe how you've attempted to connect your controllers? Make sure you've read the [usage ](https://github.com/ndeadly/MissionControl#usage) and [pairing ](https://github.com/ndeadly/MissionControl#pairing-controllers) instructions too.

The fact that none of your controllers are working sounds like user error, but it could also be unfortunate coincidence - ie. you have incompatible a counterfeit ps4 controller and your xbox controllers have either been updated to the new 5.x.x.x firmware that isn't supported or are the older non-bluetooth version.

**Comment by Master-3000** [comment]
MissionControl is a great program I would love for you to add the Halion HA-7004 Wireless Gamepad to the compatibility list.

**Reply by ndeadly** [reply]
Please fill out a controller request issue with the relevant info and I will look into it.

**Comment by Pipyakas** [comment]
hi, I'm having issues building MissionControl with latest libnx
```
bluetooth_core.cpp
D:/SWITCH/MissionControl/mc_mitm/source/bluetooth_mitm/bluetooth/bluetooth_core.cpp: In function 'ams::bluetooth::Address ams::bluetooth::core::{anonymous}::ReverseBluetoothAddress(ams::bluetooth::Address)':
D:/SWITCH/MissionControl/mc_mitm/source/bluetooth_mitm/bluetooth/bluetooth_core.cpp:39:46: error: array subscript 'uint64_t {aka long unsigned int}[0]' is partly outside array bounds of 'ams::bluetooth::Address [1]' {aka 'BtdrvAddress [1]'} [-Werror=array-bounds]
   39 |             uint64_t tmp = util::SwapEndian48(*reinterpret_cast<uint64_t *>(&address));
      |                            ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/SWITCH/MissionControl/mc_mitm/source/bluetooth_mitm/bluetooth/bluetooth_core.cpp:38:71: note: object 'address' of size 6
   38 |         bluetooth::Address ReverseBluetoothAddress(bluetooth::Address address) {
      |                                                    ~~~~~~~~~~~~~~~~~~~^~~~~~~
```

**Reply by ndeadly** [reply]
At some point the toolchain was updated to be stricter about this stuff. Build the develop branch, I made a fix for it already https://github.com/ndeadly/MissionControl/commit/4f5cb51c1e42a801ca653fde84079716b76f2454

**Reply by Pipyakas** [reply]
thanks, I've applied the fix and built it successfully! I only wanted to change some dualshock4 mapping, will custom mapping via a config file something you will look into?

**Reply by ndeadly** [reply]
>  will custom mapping via a config file something you will look into?

I haven't ruled it out for the future, but it's not something that's high priority for me. Basic button remapping can mostly be handled by the console's native remapping feature. A few fixed configuration options are planned for the future, however, like swapping d-pad and left stick, inverting stick axes etc., but allowing total freedom over remapping/reconfiguring the inputs creates a bunch of problems that I am not prepared to deal with currently.

How are you trying to remap the controls, out of interest?

**Reply by Pipyakas** [reply]
while native button remapping is present, HOS likes to remind us that we're using custom mapping everytime the controller connects, the console wakes up etc... so it annoys me a little bit.
currently I've prepared 2 builds, one with the touchpad as Plus and share/options buttons as Minus, and vice versa. If anything, using touchpad as Minus and Share as Capture would match the PS4 and give the touchpad button a much better usage imo

**Reply by ndeadly** [reply]
Ha, I didn't know that activating remapping gave that notification. I just tested it out and yeah, that's pretty annoying. I'll have a think about what it would take to provide similar basic remapping as a stop-gap.

**Reply by ndeadly** [reply]
@Pipyakas this might interest you <https://github.com/ndeadly/disable_remap_dialog>

**Comment by AlexanderLA17** [comment]
Question regarding the macros you set for all the controllers. I am using a PS5 DuelSense controller and since I love to play fighting games and frequently use select + down dpad to reset my position in training mode, I am unable to do so because it takes me to the home menu. I checked the source code for my controller, but am unable to figure out how to remove this functionality. Can you point me in the right direction? Thank you!

**Reply by ndeadly** [reply]
The button combo code is applied to all controllers in the `SwitchController` base class.

https://github.com/ndeadly/MissionControl/blob/master/mc_mitm/source/controllers/switch_controller.cpp#L242

You can either disable/change it globally here, or provide a controller-specific override for the `ApplyButtonCombos` function in the `DualsenseController`class.

**Reply by AlexanderLA17** [reply]
Thank you so much!

**Comment by ruanzhongqiang1** [comment]
Excuse me，How can the Microsoft Xbox elite wireless controller series 2 handle be used without degrading the firmware? Currently, the latest firmware of the handle cannot be found by Bluetooth。thank you。

**Reply by ndeadly** [reply]
Currently the only way to use these controllers wirelessly is to downgrade the firmware. Eventually Mission Control will support BLE connections, and then you will be able to connect controllers on any firmware

If you are unable to downgrade for whatever reason, your only options are either to connect the controller via USB instead with the help of [sys-con](https://github.com/cathery/sys-con), or wait for BLE support to be added to Mission Control.

**Reply by Domaiti** [reply]
Thanks for your support @ndeadly, right now i have the same issue, i cant make the downgrade of my Xbox One S controller. I try with my PC and with my Series S but doesnt work. I will try it one more time the next weekend with a Xbox One S. But like you said, i must wait for BLE support. Greetings from Argentina👍🏻

**Reply by ndeadly** [reply]
If you are unable to downgrade the firmware successfully, you are either doing something wrong, or trying to downgrade an incompatible controller. Only the 1708 (Xbox One S|X) and 1797 (Elite 2) models will work. The model number is printed inside the battery compartment if you are unsure.

**Reply by Domaiti** [reply]
Hey, thanks for answer me @ndeadly 😄
Actually i have the 1708 model.
In my Xbox when i press “Check controller firmware” the console show me a different message, “the controller is up to date” and i cant do the rollback.
In my PC when i type the command the app show me a little animation but doesnt appear the option to the revert. 
If you have any suggestion i will thank you a lot.

**Reply by ndeadly** [reply]
I'm not sure whether the controller can be downgraded on the Series X|S console or whether it must be an Xbox One. 

Can you show me the exact command you used on PC? I think if it's not correct the Xbox Accessories app will open as usual without the downgrade option.

**Reply by Domaiti** [reply]
I am not sure of that too, thats why i will try in a Xbox One the next friday. Probably that fix this issue (i hope).
The command that i type its “xboxaccessories:\\firmwareupdate?legacyDowngrade=true” but as you said just works for open the Xbox Accesories app as usual. I try it with the app opened before and with the app closed too.

**Reply by ndeadly** [reply]
> The command that i type its “xboxaccessories:\firmwareupdate?legacyDowngrade=true”

You need two backslashes, i.e. `xboxaccessories:\\firmwareupdate?legacyDowngrade=true`. There has been some confusion around this, as github text formatting eats the second one in some contexts. I just corrected it in the message at the top of this discussion, as it seems the corrections from the release page didn't carry over.

**Reply by Domaiti** [reply]
@ndeadly thanks for your support. With your suggestion i make the rollback and right now i make it work. Thanks buddy, i appreciate.
